### PR TITLE
Todo: list, add, update, reorder

### PR DESCRIPTION
**Pull request by AmesLexander**
I've implemented all features of todo app.

### CODE REVIEW

File: .vscode/settings.json
Comment (julycrystal) [inline .vscode/settings.json:1]: no need to commit vscode configuration.

File: api/create.sql
Comment (julycrystal) [inline api/create.sql:1]: No ending blank lines
Try best practices.

File: api/package.json
Comment (julycrystal) [inline api/package.json:1]: @types/* should be in devDependencies
no script - how do you run this project ?
by typing long commands every time?

File: api/src/controllers/index.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { Request, Response } from "express";
+import db from "../models";
+
+export const list = async (req: Request, res: Response) => {
+  try {
+    const tasks = await db.taskRepo.find();
+
+    res.status(200).json({ tasks });
+  } catch (err) {
+    res.status(404).json({ msg: "Get tasks error!" });
```
Comment (julycrystal) [inline api/src/controllers/index.ts:10]: Should use constant values from constants.ts, in one file.

File: api/tsconfig.json
Comment (julycrystal) [inline api/tsconfig.json:1]: Remove comments.
It's not worth.

File: package.json
Comment (julycrystal) [inline package.json:1]: Same issue - @types in devDependencies.

File: web/src/store/todos/index.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import axios from "axios";
+import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
+import { Task, ToDoState } from "../../utils/types";
+
+export const initialState: ToDoState = {
+  status: "success",
+};
+
+export const addTask = createAsyncThunk("todos/addTask", async (task: Task) => {
+  const response = await axios.post("http://localhost:5000/tasks", task);
```
Comment (julycrystal) [inline web/src/store/todos/index.ts:10]: same issue

File: web/src/store/todos/index.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  return response.data.task;
+});
+
+export const updateTask = createAsyncThunk(
+  "todos/updateTask",
+  async (task: Task) => {
+    const response = await axios.patch(
+      `http://localhost:5000/tasks/${task.id}`,
```
Comment (julycrystal) [inline web/src/store/todos/index.ts:18]: ...

File: web/src/store/todos/index.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        state.status = "failed";
+      })
+      .addCase(updateTask.pending, (state) => {
+        state.status = "loading";
+      })
+      .addCase(updateTask.fulfilled, (state) => {
+        state.status = "success";
+      })
+      .addCase(updateTask.rejected, (state) => {
+        state.status = "failed";
```
[the remark this answers is not in the document]
Comment (AmesLexander) [reply web/src/store/todos/index.ts:47]: How to use enums? Sorry for poor guy.

File: api/src/controllers/index.ts [hunk 5]
```diff
@@ -21,17 +29,52 @@ export const add = async (req: Request, res: Response) => {
...
+    } else {
+      await db.taskRepo.updateSort(bottom, -1);
+      array.forEach(async (_, index) => {
+        let i = bottom - index - 1;
+        await db.taskRepo.updateSort(i, i + 1);
+      });
+      await db.taskRepo.updateSort(-1, top);
+    }
+
+    res.status(200).json(UPDATE_TASKS_SUCCESS);
```
Comment (julycrystal) [inline api/src/controllers/index.ts:76]: Performance issue
Always try to minimize db access
Best if you're doing things with one statement

File: web/src/components/AddToDo/index.tsx
[the remark this answers is not in the document]
Comment (julycrystal) [reply web/src/components/AddToDo/index.tsx:1]: Not fixed.

File: web/src/components/ToDoItem/index.tsx
[the remark this answers is not in the document]
Comment (julycrystal) [reply web/src/components/ToDoItem/index.tsx:1]: Not fixed.

File: web/src/components/ToDoList/index.tsx [hunk 6]
```diff
@@ -45,68 +50,69 @@ const ToDoList = (props: ToDoListProps) => {
...
+      newTasks[end] = top;
+    } else {
+      let bottom = newTasks[end];
+      array.forEach((_, index) => {
+        let i = end - index - 1;
+        newTasks[i + 1] = newTasks[i];
+      });
+      newTasks[start] = bottom;
+    }
+    setTasks(newTasks);
```
Comment (julycrystal) [inline web/src/components/ToDoList/index.tsx:72]: Try making functions or custom hooks for code simplicity.

**Review by julycrystal — CHANGES_REQUESTED** [review]
1. You didn't fix line ending issue completely.
2. Improve code quality

**Pull request merged**